src/Interpreters/loadMetadata.cpp

```diff
@@ -310,23 +343,54 @@ void maybeConvertOrdinaryDatabaseToAtomic(ContextMutablePtr context, const Datab
         for (const auto & uuid : tables_uuids)
             DatabaseCatalog::instance().removeUUIDMappingFinally(uuid);
 
-        loadSystemDatabaseImpl(context, DatabaseCatalog::SYSTEM_DATABASE, "Atomic");
+        String path = context->getPath() + "metadata/" + escapeForFileName(database_name);
+        /// force_restore_data is needed to re-create metadata symlinks
+        loadDatabase(context, database_name, path, /* force_restore_data */ true);
+
         TablesLoader::Databases databases =
         {
-            {DatabaseCatalog::SYSTEM_DATABASE, DatabaseCatalog::instance().getSystemDatabase()},
+            {database_name, DatabaseCatalog::instance().getDatabase(database_name)},
         };
         TablesLoader loader{context, databases, /* force_restore */ true, /* force_attach */ true};
         loader.loadTables();
 
-        /// Will startup tables usual way
+        /// Startup tables if they were started before conversion and detach/attach
+        if (tables_started)
+            loader.startupTables();
     }
     catch (Exception & e)
     {
-        e.addMessage("While trying to convert {} to Atomic", database->getDatabaseName());
+        e.addMessage("Exception while trying to convert database {} from Ordinary to Atomic. It may be in some intermediate state."
+            "You can finish conversion manually by moving the rest tables from {} to {} (using RENAME TABLE)"
+            "and executing DROP DATABASE {} and RENAME DATABASE {} TO {}.",
+            database_name, database_name, tmp_name, database_name, tmp_name, database_name);
         throw;
     }
 }
 
+void maybeConvertSystemDatabase(ContextMutablePtr context)
+{
+    /// TODO remove this check, convert system database unconditionally
+    if (context->getSettingsRef().allow_deprecated_database_ordinary)
+        return;
+
+    maybeConvertOrdinaryDatabaseToAtomic(context, DatabaseCatalog::SYSTEM_DATABASE, /* tables_started */ false);
+}
+
+void convertDatabasesEnginesIfNeed(ContextMutablePtr context)
+{
+    auto convert_flag_path = fs::path(context->getFlagsPath()) / "convert_ordinary_to_atomic";
+    if (!fs::exists(convert_flag_path))
+        return;
+
+    LOG_INFO(&Poco::Logger::get("loadMetadata"), "Found convert_ordinary_to_atomic file in flags directory, "
+                                                 "will try to convert all Ordinary databases to Atomic");
+    fs::remove(convert_flag_path);
+
+    for (const auto & [name, _] : DatabaseCatalog::instance().getDatabases())
+        if (name != DatabaseCatalog::SYSTEM_DATABASE)
+            maybeConvertOrdinaryDatabaseToAtomic(context, name, /* tables_started */ true);
```

Comment:
I am wondering if it can be parallelised and whether this parallelisation is needed... Seems like not so much work, but definitely not for this PR

Reply:
I think it's not worth it, each ClickHouse server will execute this code at most once. And moving tables should work quite fast